File: Python/InvIQ.py
import random
import time

def program_greeting():
    print("\nHello and welcome to my Inventory program!")
    print("Author: Jashandeep Singh")
    print("Due Date: October 22, 2023")

class Date:
    def __init__(self):
        self.year_of_manufacture = time.localtime().tm_year

    def get_current_year(self):
        return self.year_of_manufacture

    def check_year(self):
        system_date = Date()
        if system_date.get_current_year() == 2023:
            print("Date checkYear: Pass - The current year is 2023.")
        else:
            print("Date checkYear: Fail - The year is not 2023.")

class RandNo:
    has_been_seeded = False

    def __init__(self):
        if not RandNo.has_been_seeded:
            random.seed()
            RandNo.has_been_seeded = True

    def get_random(self, minimum, maximum):
        return random.uniform(minimum, maximum)

class InVal:
    @staticmethod
    def item_id_validation():
        product_id = input("Enter Product ID (5 digits only): ")
        return product_id

    @staticmethod
    def stock_amount_validation():
        stock_level = int(input("Enter Stock Level (non-negative): "))
        return stock_level

    @staticmethod
    def cost_validation():
        base_price = float(input("Enter Base Price (non-negative): "))
        return base_price

class StockItem:
    def __init__(self, product_code, stock_quantity, base_cost, sale_price, entry_date):
        self.product_code = product_code
        self.stock_quantity = stock_quantity
        self.base_cost = base_cost
        self.sale_price = sale_price
        self.entry_date = entry_date

    def __str__(self):
        return (f"Product Code: {self.product_code}\n"
                f"Stock Quantity: {self.stock_quantity}\n"
                f"Base Cost: ${self.base_cost:.2f}\n"
                f"Sale Price: ${self.sale_price:.2f}\n"
                f"Entry Date: {self.entry_date.get_current_year()}\n")

class StockControl:
    def __init__(self):
        self.stock_list = []

    def add_stock_item(self, new_item):
        self.stock_list.append(new_item)

    def remove_stock_item(self):
        if self.stock_list:
            self.stock_list.pop()

    def update_stock_item(self, code, quantity, cost):
        for item in self.stock_list:
            if item.product_code == code:
                item.stock_quantity = quantity
                item.base_cost = cost
                item.sale_price = cost * 1.25
                break

    def show_stock_items(self):
        for item in self.stock_list:
            print(item)

    def get_list_size(self):
        return len(self.stock_list)

def stock_menu(inventory_control, random_no_gen):
    while True:
        print("\nInventory Management Menu")
        print("A - Add Inventory")
        print("D - Delete Inventory")
        print("E - Edit Inventory")
        print("P - Print Inventory")
        print("Q - Quit Program")
        user_choice = input("Enter choice: ").upper()

        if user_choice == 'A':
            product_code = InVal.item_id_validation()
            stock_quantity = InVal.stock_amount_validation()
            base_cost = InVal.cost_validation()
            sale_price = random_no_gen.get_random(base_cost, base_cost * 1.5)
            entry_date = Date()
            new_item = StockItem(product_code, stock_quantity, base_cost, sale_price, entry_date)
            inventory_control.add_stock_item(new_item)
        elif user_choice == 'D':
            inventory_control.remove_stock_item()
        elif user_choice == 'E':
            if inventory_control.get_list_size() > 0:
                code_to_update = InVal.item_id_validation()
                updated_quantity = InVal.stock_amount_validation()
                updated_cost = InVal.cost_validation()
                inventory_control.update_stock_item(code_to_update, updated_quantity, updated_cost)
            else:
                print("\nNo inventory items to edit.")
        elif user_choice == 'P':
            inventory_control.show_stock_items()
        elif user_choice == 'Q':
            print("Quitting the program.")
            break
        else:
            print("Invalid selection. Please try again.")

def unit_test():
    Date().check_year()

def main():
    program_greeting()
    unit_test()
    inventory_control = StockControl()
    random_no_gen = RandNo()
    stock_menu(inventory_control, random_no_gen)

main()

File: CPP/InvIQ.cpp
//Invinq.cpp
//Jashandeep Singh, CISP 400 
// 10/22/2023
#include <iostream>                                 
#include <string>                                      
#include <iomanip>                                     
#include <cstdlib>                                     
#include <ctime>
using namespace std;

void programGreeting(){
  cout << "\nHello and welcome to my Inventory program!" << endl;
  cout << "Author: Jashandeep Singh" << endl;
  cout << "Due Date: October 22, 2023" << endl;
}

// Specification B1 - Date class
class Date {
private:
    int yearOfManufacture;

public:
    Date() {
        time_t timer = time(nullptr);
        tm* timeStruct = localtime(&timer);
        yearOfManufacture = timeStruct->tm_year + 1900;
    }

    int getCurrentYear() const {
        return yearOfManufacture;
    }

    void checkYear() {
        Date systemDate;
        if (systemDate.getCurrentYear() == 2023) {
            cout << "Date checkYear: Pass - The current year is 2023." << endl;
        } else {
            cout << "Date checkYear: Fail - The year is not 2023." << endl;
        }
    }
};

// Specification B2 - RandNo Class
// Specification B3 - Singleton Design Pattern
class RandNo {
private:
    static bool hasBeenSeeded;

public:
    RandNo() {
        if (!hasBeenSeeded) {
            srand(static_cast<unsigned int>(time(nullptr)));
            hasBeenSeeded = true;
        }
    }

    double getRandom(double minimum, double maximum) {
        return minimum + static_cast<double>(rand()) / (RAND_MAX / (maximum - minimum));
    }

    // Specification A3 - Overload operator>>
    friend istream& operator>>(istream& input, RandNo& randomNumberGenerator) {
        int seedValue;
        cout << "\nEnter a custom seed for RNG: ";
        input >> seedValue;
        srand(seedValue);
        hasBeenSeeded = true;
        return input;
    }
};

bool RandNo::hasBeenSeeded = false; 

// Specification B4 - Inventory Entry Input Validation
class InVal {
public:
    static string itemIDValidation() {
        string productID;
        cout << "Enter Product ID (5 digits only): ";
        cin >> productID;
        return productID; // Simple validation for example purposes
    }

    static int stockAmountValidation() {
        int stockLevel;
        cout << "Enter Stock Level (non-negative): ";
        cin >> stockLevel;
        return stockLevel; // Simple validation for example purposes
    }

    static double costValidation() {
        double basePrice;
        cout << "Enter Base Price (non-negative): ";
        cin >> basePrice;
        return basePrice; // Simple validation for example purposes
    }
};

// Specification A2 - Overload operator<<
struct StockItem {
    string productCode;
    int stockQuantity;
    double baseCost;
    double salePrice;
    Date entryDate;

    friend ostream& operator<<(ostream& outputStream, const StockItem& item) {
        outputStream << "Product Code: " << item.productCode << "\n"
                     << "Stock Quantity: " << item.stockQuantity << "\n"
                     << "Base Cost: $" << fixed << setprecision(2) << item.baseCost << "\n"
                     << "Sale Price: $" << fixed << setprecision(2) << item.salePrice << "\n"
                     << "Entry Date: " << item.entryDate.getCurrentYear() << "\n";
        return outputStream;
    }
};

// Specification C2 - Dynamic Array  
// Specification C3 - Resize Array
class StockControl {
private:
    StockItem* stockList;
    int listSize;

public:
    StockControl() : listSize(0), stockList(new StockItem[1]) {}

    ~StockControl() {
        delete[] stockList;
    }

    void addStockItem(const StockItem& newItem) {
        StockItem* pTmp = new StockItem[listSize + 1];
        for (int index = 0; index < listSize; ++index) {
              pTmp[index] = stockList[index];
        }
          pTmp[listSize] = newItem;
        delete[] stockList;
        stockList = pTmp;
        ++listSize;
    }

    void removeStockItem() {
        if (listSize > 0) {
            StockItem* pTmp = new StockItem[listSize - 1];
            for (int index = 0; index < listSize - 1; ++index) {
                  pTmp[index] = stockList[index];
            }
            delete[] stockList;
            stockList = pTmp;
            --listSize;
        }
    }

    // Specification A1 - Edit Inventory
    void updateStockItem(const string& code, int quantity, double cost) {
        for (int index = 0; index < listSize; ++index) {
            if (stockList[index].productCode == code) {
                stockList[index].stockQuantity = quantity;
                stockList[index].baseCost = cost;
                stockList[index].salePrice = cost * 1.25; 
                return;
            }
        }
    }

    void showStockItems() const {
        for (int index = 0; index < listSize; ++index) {
            cout << stockList[index] << endl;
        }
    }

    int getListSize() const {
        return listSize;
    }
};

// Specification C1 - Alpha Menu 
// Specification C4 - Menu Input Validation
void stockMenu(StockControl &inventoryControl, RandNo &randomNoGen) {
    char userChoice;
    do {
        cout << "\nInventory Management Menu" << endl;
        cout << "A - Add Inventory" << endl;
        cout << "D - Delete Inventory" << endl;
        cout << "E - Edit Inventory" << endl;
        cout << "P - Print Inventory" << endl;
        cout << "Q - Quit Program" << endl;
        cout << "Enter choice: ";
        cin >> userChoice;
        userChoice = toupper(userChoice);

        switch (userChoice) {
            case 'A': {
                // Add Inventory
                StockItem newItem;
                newItem.productCode = InVal::itemIDValidation();
                newItem.stockQuantity = InVal::stockAmountValidation();
                newItem.baseCost = InVal::costValidation();
                newItem.salePrice = randomNoGen.getRandom(newItem.baseCost, newItem.baseCost * 1.5);
                newItem.entryDate = Date();
                inventoryControl.addStockItem(newItem);
                break;
            }
            case 'D': {
                // Delete Inventory
                inventoryControl.removeStockItem();
                break;
            }
            case 'E': {
                // Edit Inventory
                if (inventoryControl.getListSize() > 0) {
                    string codeToUpdate = InVal::itemIDValidation();
                    int updatedQuantity = InVal::stockAmountValidation();
                    double updatedCost = InVal::costValidation();
                    inventoryControl.updateStockItem(codeToUpdate, updatedQuantity, updatedCost);
                } else {
                    cout << "\nNo inventory items to edit." << endl;
                }
                break;
            }
            case 'P': {
                // Print Inventory
                inventoryControl.showStockItems();
                break;
            }
            case 'Q': {
                // Quit Program
                cout << "Quitting the program." << endl;
                return;
            }
            default: {
                cout << "Invalid selection. Please try again." << endl;
                break;
            }
        }
    } while (userChoice != 'Q');
}

// Specification A4 - UnitTest() function in main()
void UnitTest() {
    // Unit tests for Date and RandNo classes
    Date().checkYear();
}

int main() {
    programGreeting();
    UnitTest();
    StockControl inventoryControl;
    RandNo randomNoGen;
    stockMenu(inventoryControl, randomNoGen);
    return 0;
}